<a href="https://colab.research.google.com/github/damiannogueira/flyrank-ml-intership/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/damiannogueira/flyrank-ml-intership/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)



## 1. My lane (or freestyle) and why

### Lane: Refresh / Content Opportunity Scoring

I chose this lane because it focuses on a clear practical decision: which pages should be reviewed first. It also connects well with the starter notebooks, where I already worked with declining pages, ranking rules, Precision@K, and simple decision-tree models. This lane will let me explore whether observable search and content signals can help prioritize content review in a useful and explainable way.

## 2. The question: decision, action, cost of a wrong call

### Research question

Which pages should be prioritized for content review based on observable search and content signals?

### Decision

The decision is which pages should be reviewed first when there is limited time to analyze or update content.

### Action

A content or SEO team could use the ranked output to decide which pages to review, refresh, expand, protect, prune, or monitor first.

### Cost of a wrong call

A wrong recommendation could waste time on a page that does not need attention, while a page with a real opportunity or decline could be missed. This makes the quality of the ranking important, especially for the pages placed at the top.

## 3. Quick look at the data (2-3 real numbers)

I will use the starter dataset to check whether there are enough pages with declining performance and enough stale but visible pages to make content-review prioritization a useful problem to explore.

In [1]:
# Setup for Colab:
# clone my repository if needed and move to the repo root
# so the notebook can access the starter dataset with relative paths.

import os
import sys
import subprocess

REPO_URL = "https://github.com/damiannogueira/flyrank-ml-intership"
REPO_DIR = "flyrank-ml-intership"

if "google.colab" in sys.modules:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(
            ["git", "clone", "--depth", "1", REPO_URL, REPO_DIR],
            check=True
        )

    os.chdir(REPO_DIR)

print("Working dir:", os.getcwd())

Working dir: /content/flyrank-ml-intership


In [2]:
# Load the starter dataset and calculate a few numbers
# that help evaluate whether this lane is worth exploring.

import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

total_pages = len(df)

declining_pages = (
    df["trend_direction"].str.lower() == "down"
).sum()

declining_pct = declining_pages / total_pages * 100

stale_visible_pages = (
    (df["days_since_last_update"] >= 180) &
    (df["impressions_90d"] >= 500)
).sum()

print(f"Total pages: {total_pages:,}")
print(f"Declining pages: {declining_pages:,} ({declining_pct:.1f}%)")
print(f"Stale and visible pages: {stale_visible_pages:,}")

Total pages: 30,000
Declining pages: 16,262 (54.2%)
Stale and visible pages: 17


### Observation

The starter dataset contains 30,000 pages, and 16,262 of them (54.2%) are currently labeled as declining. This suggests that there is a large enough group of pages with negative movement to make prioritization relevant.

Using the simple rule of at least 180 days since the last update and at least 500 impressions in the last 90 days, only 17 pages are classified as both stale and visible. This shows that a single strict rule captures only a small subset of the available pages.

These numbers make the Refresh / Content Opportunity Scoring lane worth exploring because the problem is not only finding pages with declining performance, but deciding which ones deserve attention first using multiple observable signals.

## 4. Careful words: what I can and can't claim

### What I can claim

This project can describe observed patterns in the data and support decisions about which pages may deserve review first. It can compare ranking methods, identify useful signals, and show whether one prioritization approach performs better than a simple baseline on the chosen evaluation metrics.

### What I can't claim

I cannot claim that a content refresh will cause a page to recover, that any feature is a confirmed Google ranking factor, or that the model predicts Google's algorithm. The results will be observational and should be used as decision-support, with human review before taking action.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.